qemu/tests/win_virtio_driver_installer_pretest.py

```diff
@@ -0,0 +1,177 @@
+    3) Clear the installed drivers/agent/virtiofs.
+    4) Install/update driver via virtio-win-guest-tools.exe.
+    5) if creating virtio-win iso
+        a. copy driver folder to host
+        b. create a virtio-win iso including driver folders
+    6) if test installer
+        a. Run virtio-win-guest-tools.exe signature check command in guest.
+        b. Verify the qemu-ga version match expected version.
+        c. Run driver signature check command in guest.
+           Verify target driver.
+        d) service check.
+    7) uninstall test if needed
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+    def change_virtio_media(cdrom_virtio):
+        """
+        change iso for virtio-win
+        :param cdrom_virtio: iso file
+        """
+        virtio_iso = utils_misc.get_path(data_dir.get_data_dir(),
+                                         cdrom_virtio)
+        test.log.info("Changing virtio iso image to '%s'", virtio_iso)
+        vm.change_media("drive_virtio", virtio_iso)
+
+    # gagent version check test config
+    qemu_ga_pkg = params["qemu_ga_pkg"]
+    gagent_pkg_info_cmd = params["gagent_pkg_info_cmd"]
+    gagent_install_cmd = params["gagent_install_cmd"]
+    gagent_uninstall_cmd = params["gagent_uninstall_cmd"]
+
+    run_install_cmd = params["run_install_cmd"]
+    installer_pkg_check_cmd = params["installer_pkg_check_cmd"]
+    driver_test_params = params.get("driver_test_params", "{}")
+    driver_test_params = ast.literal_eval(driver_test_params)
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+
+    # get expected agent version
+    expected_gagent_version = win_driver_installer_test.install_gagent(
+        session, test,
+        qemu_ga_pkg,
+        gagent_install_cmd,
+        gagent_pkg_info_cmd)
+
+    win_driver_installer_test.uninstall_gagent(session, test,
+                                               gagent_uninstall_cmd)
+    # delete virtiofs service
+    virtio_fs_utils.delete_viofs_serivce(test, params, session)
+
+    # uninstall the previous drivers
+    win_driver_installer_test.win_uninstall_all_drivers(session,
+                                                        test, params)
+    session = vm.reboot(session)
+
+    if params.get("update_test"):
+        error_context.context("Install drivers from previous installer",
+                              test.log.info)
+        change_virtio_media(params["cdrom_virtio_downgrade"])
+        # add sleep time in case iso changed windows
+        # pop up before the installer windows
+        time.sleep(2)
+        win_driver_installer_test.install_test_with_screen_on_desktop(
+            vm, session, test,
+            run_install_cmd,
+            installer_pkg_check_cmd,
+            copy_files_params=params)
+        error_context.context("Start viofs service...")
+        params["is_pretest"] = None
```

Comment:
Because the folder structure is different between the common virtio-win iso with virtio-win-pretest iso.
such as: viofs/ws2022/viofs.exe vs viofs/viofs.exe
Here we changed iso to change_virtio_media(params["cdrom_virtio_downgrade"])